ConnectorActionableMsgs/02 Cards and Actions/Demos/03-Expense-Approval/readme.md

```diff
@@ -0,0 +1,319 @@
+            HttpResponseMessage response;
+            var approvalResult = default(string);
+
+            if(action == "approve")
+            {
+                approvalResult = "Approved";
+                response = ActionableMessageHelper.CreateCardResponse(Request, HttpStatusCode.OK, "The expense was approved.");
+                Trace.TraceInformation("Expense report approved: " + value);
+            }
+            else
+            {
+                approvalResult = "Declined";
+                response = ActionableMessageHelper.CreateCardResponse(Request, HttpStatusCode.OK, "The expense was declined.");
+                Trace.TraceInformation("Expense report declined: " + value);
+            }
+
+            //Refresh the card
+            response.Headers.Add("CARD-UPDATE-IN-BODY", "true");
+            string refreshCard = ActionableMessageHelper.GetCardBody(value, approvalResult, ret.ValidationResult.ActionPerformer);            
+            response.Content = new StringContent(refreshCard);
+
+            return response;
+        }
+    }
+}
+````
+### Publish the Azure Web Application
+The Web API that you just created will be called from Microsoft, so it needs to be available publicly and not running locally on localhost. Right-click the web application project and choose **Publish**. Choose **Select Existing** and click **OK**. 
+
+![](../../Images/publish.png)
+
+Choose your existing Web App and click **OK**.
+
+In the **Publish** window, click the **Settings...** link. Click the **Settings** tab and change the configuration to **Debug**. 
+
+![](../../Images/webdeploy.png)
+
+**Save** your changes then click **Publish** to publish your web application code.
+
+### Debug the Azure Web App
+You can attach a debugger to an Azure Web App similar to how you attach a debugger to a local process during debugging. In Visual Studio 2017, open the **Cloud Explorer** pane and expand the **App Services** node to show your Azure Web App. Right-click your web app and choose **Attach Debugger**.
+
+![](../../Images/debugger.png)
+
+Set a breakpoint in the Web API controller to see when messages arrive and debug interactively.
+
+### Test the card
+In the previous section, you sent a card to yourself using both PowerShell and a webhook. Those email messages should still be in your inbox (if not, repeat the previous section exercise). Open the email and click the **Approve** button. Provide text, simulating comments to an approval form, and click **Submit**. 
+
+![](../../Images/actioncard.png)
+
+The debugger in your code is reached, and you can step through the code to see the bearer token is validated, the sender and email domains are validated, the refresh card body is retrieved and the response is sent with the appropriate headers.
+
+In your email client, the card is now updated to reflect the data sent in the refresh card.
+
+![](../../Images/refreshcard.png)
```

Comment:
It might be a good idea to remove the "Please approve this expense report" line from the refresh card. Not a deal-breaker, but logically it makes sense.